src/Pate/Verification/PairGraph.hs

```diff
@@ -0,0 +1,878 @@
+  PPa.PatchPair (PB.ConcreteBlock arch) {- ^ next entry point -} ->
+  PPa.PatchPair (PB.ConcreteBlock arch) {- ^ return point -} ->
+  EquivM sym arch (PairGraph sym arch)
+handleOrdinaryFunCall bundle currBlock d gr pPair _pRetPair =
+  -- TODO? we are just ignoring the retPair...
+  handleJump bundle currBlock d gr pPair
+
+handleJump ::
+  SimBundle sym arch ->
+  PPa.BlockPair arch {- ^ current entry point -} ->
+  AbstractDomain sym arch {- ^ current abstract domain -} ->
+  PairGraph sym arch ->
+  PPa.BlockPair arch {- ^ next entry point -} ->
+  EquivM sym arch (PairGraph sym arch)
+handleJump bundle currBlock d gr pPair =
+  do d' <- getTargetDomain pPair gr
+     md <- widenPostcondition bundle d d'
+     case md of
+       NoWideningRequired ->
+         do traceBundle bundle "Did not need to widen"
+            return gr
+       WideningError msg ->
+         do traceBundle bundle ("Error during widening: " ++ msg)
+            return gr -- TODO? better error handling?
+       Widen _ d'' ->
+         case updateDomain gr currBlock pPair d'' of
+           Nothing ->
+             do traceBundle bundle ("Ran out of gas while widening postconditon! " ++ show currBlock ++ " " ++ show pPair)
+                return gr -- TODO? better error handling?
+           Just gr' ->
+             do traceBundle bundle "Successfully widened postcondition"
+                return gr'
+
+getTargetDomain ::
+  PPa.BlockPair arch ->
+  PairGraph sym arch ->
+  EquivM sym arch (AbstractDomain sym arch)
+getTargetDomain pPair gr =
+  case Map.lookup pPair (pairGraphDomains gr) of
+    Just d  -> return d
+    Nothing ->
+      -- initial state of the pair graph: choose the universal domain that equates as much as possible
+      PVD.universalDomainSpec pPair
+
+data WidenLocs sym arch =
+  WidenLocs
+    (Set (Some (MM.ArchReg arch)))
+    (Set (Some (PMc.MemCell sym arch)))
+
+instance PA.ValidArch arch => Show (WidenLocs sym arch) where
+  show (WidenLocs regs cells) =
+    unlines [ unwords (map show (Set.toList regs))
+            , show (Set.size cells) ++ " memory locations"
+            ]
+
+instance (OrdF (W4.SymExpr sym), PA.ValidArch arch) => Semigroup (WidenLocs sym arch) where
+  (WidenLocs r1 m1) <> (WidenLocs r2 m2) = WidenLocs (r1 <> r2) (m1 <> m2)
+
+instance (OrdF (W4.SymExpr sym), PA.ValidArch arch) => Monoid (WidenLocs sym arch) where
+  mempty = WidenLocs mempty mempty
+
+data WidenResult sym arch
+  = NoWideningRequired
+  | WideningError String
+  | Widen (WidenLocs sym arch) (AbstractDomain sym arch)
+
+-- | Try the given widening stratetigs s one at a time,
+--   until the first one that computes some nontrival
+--   widening, or returns an error.
+tryWidenings ::
+  [IO (WidenResult sym arch)] ->
+  IO (WidenResult sym arch)
+tryWidenings [] = return NoWideningRequired
+tryWidenings (x:xs) =
+  x >>= \case
+    NoWideningRequired -> tryWidenings xs
+    res -> return res
+
+widenPostcondition :: forall sym arch.
```

Comment:
This is probably fine for now, but this is probably one of the things that needs the most documentation once we are happier with it

Reply:
Agreed